### PR TITLE
Add date and date_nanos conversion to the numeric_type sort option

### PR DESCRIPTION
**Pull request by jimczi**
This change adds an option to convert a `date` field to nanoseconds resolution
 and a `date_nanos` field to millisecond resolution when sorting.
The resolution of the sort can be set using the `numeric_type` option of the
field sort builder. The conversion is done at the shard level and is restricted
to dates from 1970 to 2262 for the nanoseconds resolution in order to avoid
numeric overflow.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/time/DateUtils.java [hunk 1]
```diff
@@ -129,6 +129,27 @@ public static Instant toInstant(long nanoSecondsSinceEpoch) {
...
+            throw new IllegalArgumentException("milliSeconds [" + milliSecondsSinceEpoch + "] are before the epoch in 1970 and cannot " +
+                "be converted to nanoseconds");
+        } else if (milliSecondsSinceEpoch > MAX_NANOSECOND_INSTANT.toEpochMilli()) {
+            throw new IllegalArgumentException("milliSeconds [" + milliSecondsSinceEpoch + "] are after 2262-04-11T23:47:16.854775807 " +
+                "and cannot be converted to nanoseconds");
+        }
+
+        if (milliSecondsSinceEpoch == 0) {
+            return 0;
+        }
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/time/DateUtils.java]: why this special case?

File: server/src/main/java/org/elasticsearch/common/time/DateUtils.java [hunk 2]
```diff
@@ -129,6 +129,27 @@ public static Instant toInstant(long nanoSecondsSinceEpoch) {
...
+     * Convert a nanosecond timestamp in milliseconds
+     *
+     * @param milliSecondsSinceEpoch the millisecond since the epoch
+     * @return                      the nanoseconds since the epoch
+     */
+    public static long toNanoSeconds(long milliSecondsSinceEpoch) {
+        if (milliSecondsSinceEpoch < 0) {
+            throw new IllegalArgumentException("milliSeconds [" + milliSecondsSinceEpoch + "] are before the epoch in 1970 and cannot " +
+                "be converted to nanoseconds");
+        } else if (milliSecondsSinceEpoch > MAX_NANOSECOND_INSTANT.toEpochMilli()) {
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/time/DateUtils.java]: how costly is the toEpochMillis call?

**Review by jpountz — APPROVED** [review]
One question, other than that LGTM.

File: server/src/main/java/org/elasticsearch/common/time/DateUtils.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/common/time/DateUtils.java]: just a copy/paste from the other conversion, this is not needed

File: server/src/main/java/org/elasticsearch/common/time/DateUtils.java (hunk 2 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/common/time/DateUtils.java]: Yes it's costly to call it on every value, I pushed https://github.com/elastic/elasticsearch/pull/40199/commits/b0e2a2007330ecadacce0e2a1ccfa623f69b7717 to compute this value once.

**Review by jpountz — APPROVED** [review]

**Pull request merged**